File: implementations/c/provekit-self-contracts/src/jcs.c
/* SPDX-License-Identifier: Apache-2.0 */
/*
 * Generic JSON Value tree + JCS canonicalizer (RFC 8785).
 *
 * Rules:
 *   - Object keys sorted by Unicode code-point order. For ASCII-only
 *     keys (which is everything the protocol uses) this collapses to
 *     bytewise strcmp.
 *   - Strings: UTF-8 verbatim (>= U+0080 emitted as raw multi-byte
 *     sequences). Escape `"` and `\` and U+0000..U+001F as `\u00XX`
 *     with lowercase hex.
 *   - Integers: plain decimal (i64).
 *   - true / false / null verbatim.
 *   - No whitespace anywhere.
 *
 * Mirrors implementations/rust/provekit-canonicalizer/src/{value,jcs}.rs
 * and the IR-coupled provekit-ir/src/jcs.c (this file is the generic
 * value-tree variant; the IR variant in provekit-ir handles IR nodes
 * directly without an intermediate Value tree).
 */

#include "provekit/self_contracts.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* --- Value constructors ------------------------------------------------- */

static pksc_value *alloc_value(pksc_value_kind k) {
    pksc_value *v = (pksc_value *)calloc(1, sizeof(*v));
    if (v) v->kind = k;
    return v;
}

static char *dup_str(const char *s) {
    if (!s) return NULL;
    size_t n = strlen(s);
    char *p = (char *)malloc(n + 1);
    if (!p) return NULL;
    memcpy(p, s, n + 1);
    return p;
}

pksc_value *pksc_v_null(void) { return alloc_value(PKSC_V_NULL); }

pksc_value *pksc_v_bool(int b) {
    pksc_value *v = alloc_value(PKSC_V_BOOL);
    if (v) v->as.b = b ? 1 : 0;
    return v;
}

pksc_value *pksc_v_int(int64_t i) {
    pksc_value *v = alloc_value(PKSC_V_INT);
    if (v) v->as.i = i;
    return v;
}

pksc_value *pksc_v_str(const char *s) {
    pksc_value *v = alloc_value(PKSC_V_STR);
    if (!v) return NULL;
    v->as.s = dup_str(s ? s : "");
    if (!v->as.s) { free(v); return NULL; }
    return v;
}

pksc_value *pksc_v_arr_new(void) {
    pksc_value *v = alloc_value(PKSC_V_ARR);
    return v;
}

pksc_value *pksc_v_obj_new(void) {
    pksc_value *v = alloc_value(PKSC_V_OBJ);
    return v;
}

int pksc_v_arr_push(pksc_value *arr, pksc_value *item) {
    if (!arr || arr->kind != PKSC_V_ARR || !item) return -1;
    size_t n = arr->as.arr.n;
    pksc_value **p = (pksc_value **)realloc(arr->as.arr.items, (n + 1) * sizeof(*p));
    if (!p) return -1;
    arr->as.arr.items = p;
    p[n] = item;
    arr->as.arr.n = n + 1;
    return 0;
}

int pksc_v_obj_set(pksc_value *obj, const char *key, pksc_value *value) {
    if (!obj || obj->kind != PKSC_V_OBJ || !key || !value) return -1;
    /* No dedupe — caller is responsible. JCS sorts at emit anyway, and
     * duplicate keys would render the document invalid JSON regardless. */
    if (obj->as.obj.n == obj->as.obj.cap) {
        size_t new_cap = obj->as.obj.cap ? obj->as.obj.cap * 2 : 4;
        pksc_kv *p = (pksc_kv *)realloc(obj->as.obj.entries, new_cap * sizeof(*p));
        if (!p) return -1;
        obj->as.obj.entries = p;
        obj->as.obj.cap = new_cap;
    }
    pksc_kv *e = &obj->as.obj.entries[obj->as.obj.n];
    e->key = dup_str(key);
    if (!e->key) return -1;
    e->value = value;
    obj->as.obj.n += 1;
    return 0;
}

void pksc_value_free(pksc_value *v) {
    if (!v) return;
    switch (v->kind) {
        case PKSC_V_NULL:
        case PKSC_V_BOOL:
        case PKSC_V_INT:
            break;
        case PKSC_V_STR:
            free(v->as.s);
            break;
        case PKSC_V_ARR:
            for (size_t i = 0; i < v->as.arr.n; i++) {
                pksc_value_free(v->as.arr.items[i]);
            }
            free(v->as.arr.items);
            break;
        case PKSC_V_OBJ:
            for (size_t i = 0; i < v->as.obj.n; i++) {
                free(v->as.obj.entries[i].key);
                pksc_value_free(v->as.obj.entries[i].value);
            }
            free(v->as.obj.entries);
            break;
    }
    free(v);
}

/* --- JCS encoder -------------------------------------------------------- */

static int encode_value(pksc_bytes *out, const pksc_value *v);

static int encode_string(pksc_bytes *out, const char *s) {
    if (pksc_bytes_append_byte(out, '"') != 0) return -1;
    /* Iterate raw bytes. UTF-8 multi-byte continuation bytes always have
     * the high bit set, so they never collide with the ASCII control-char
     * range (< 0x20) we escape — emit them verbatim and the receiver
     * sees the same UTF-8 bytes the producer wrote. Matches the Rust
     * encoder's `c.chars()` shape because every char encodes to its UTF-8
     * bytes when pushed to a String. */
    for (const unsigned char *p = (const unsigned char *)s; *p; p++) {
        unsigned char c = *p;
        if (c == '"') {
            if (pksc_bytes_append(out, (const uint8_t *)"\\\"", 2) != 0) return -1;
        } else if (c == '\\') {
            if (pksc_bytes_append(out, (const uint8_t *)"\\\\", 2) != 0) return -1;
        } else if (c < 0x20) {
            char esc[7];
            int n = snprintf(esc, sizeof(esc), "\\u00%02x", c);
            if (n < 0) return -1;
            if (pksc_bytes_append(out, (const uint8_t *)esc, (size_t)n) != 0) return -1;
        } else {
            if (pksc_bytes_append_byte(out, c) != 0) return -1;
        }
    }
    return pksc_bytes_append_byte(out, '"');
}

/* Sort key indexes so we can iterate sorted without mutating the original. */
typedef struct {
    const pksc_kv *entries;
    size_t        *indexes;
    size_t         n;
} sort_ctx;

static int kv_compare(const void *pa, const void *pb, const pksc_kv *entries) {
    size_t ia = *(const size_t *)pa;
    size_t ib = *(const size_t *)pb;
    return strcmp(entries[ia].key, entries[ib].key);
}

/* qsort_r is platform-divergent; use a thread-local context. We don't
 * need thread safety inside a single encode call, so a static is fine
 * for serial usage. JCS encoding is leaf-only inside one call. */
static const pksc_kv *g_sort_entries;

static int kv_compare_static(const void *pa, const void *pb) {
    return kv_compare(pa, pb, g_sort_entries);
}

static int encode_object(pksc_bytes *out, const pksc_value *v) {
    if (pksc_bytes_append_byte(out, '{') != 0) return -1;
    size_t n = v->as.obj.n;
    if (n > 0) {
        size_t *idx = (size_t *)malloc(n * sizeof(*idx));
        if (!idx) return -1;
        for (size_t i = 0; i < n; i++) idx[i] = i;
        g_sort_entries = v->as.obj.entries;
        qsort(idx, n, sizeof(*idx), kv_compare_static);
        g_sort_entries = NULL;

        for (size_t i = 0; i < n; i++) {
            if (i > 0) {
                if (pksc_bytes_append_byte(out, ',') != 0) { free(idx); return -1; }
            }
            const pksc_kv *e = &v->as.obj.entries[idx[i]];
            if (encode_string(out, e->key) != 0) { free(idx); return -1; }
            if (pksc_bytes_append_byte(out, ':') != 0) { free(idx); return -1; }
            if (encode_value(out, e->value) != 0) { free(idx); return -1; }
        }
        free(idx);
    }
    return pksc_bytes_append_byte(out, '}');
}

static int encode_array(pksc_bytes *out, const pksc_value *v) {
    if (pksc_bytes_append_byte(out, '[') != 0) return -1;
    for (size_t i = 0; i < v->as.arr.n; i++) {
        if (i > 0) {
            if (pksc_bytes_append_byte(out, ',') != 0) return -1;
        }
        if (encode_value(out, v->as.arr.items[i]) != 0) return -1;
    }
    return pksc_bytes_append_byte(out, ']');
}

static int encode_value(pksc_bytes *out, const pksc_value *v) {
    if (!v) return -1;
    switch (v->kind) {
        case PKSC_V_NULL:
            return pksc_bytes_append(out, (const uint8_t *)"null", 4);
        case PKSC_V_BOOL:
            return v->as.b
                ? pksc_bytes_append(out, (const uint8_t *)"true", 4)
                : pksc_bytes_append(out, (const uint8_t *)"false", 5);
        case PKSC_V_INT: {
            char buf[32];
            int  n = snprintf(buf, sizeof(buf), "%lld", (long long)v->as.i);
            if (n < 0) return -1;
            return pksc_bytes_append(out, (const uint8_t *)buf, (size_t)n);
        }
        case PKSC_V_STR:
            return encode_string(out, v->as.s);
        case PKSC_V_ARR:
            return encode_array(out, v);
        case PKSC_V_OBJ:
            return encode_object(out, v);
    }
    return -1;
}

int pksc_jcs_encode(pksc_bytes *out, const pksc_value *v) {
    if (!out) return -1;
    return encode_value(out, v);
}

char *pksc_jcs_encode_string(const pksc_value *v) {
    pksc_bytes b;
    pksc_bytes_init(&b);
    if (pksc_jcs_encode(&b, v) != 0) {
        pksc_bytes_free(&b);
        return NULL;
    }
    /* NUL-terminate for caller convenience. */
    if (pksc_bytes_append_byte(&b, '\0') != 0) {
        pksc_bytes_free(&b);
        return NULL;
    }
    return (char *)b.data;
}

File: implementations/zig/provekit-self-contracts/src/jcs.zig
// SPDX-License-Identifier: Apache-2.0
//
// JCS-JSON encoder (RFC 8785) + Value tree.
//
// Mirrors implementations/rust/provekit-canonicalizer/src/{jcs.rs,value.rs}
// 1:1. The same rules apply across all 11 kits:
//
//   - Object keys sorted by Unicode code-point order. For ASCII-only keys
//     this collapses to byte-order; the protocol's keys are all ASCII so
//     byte-order suffices. Cross-kit byte-equivalence depends on this.
//   - Numbers: integers serialized as plain decimal digits (we only carry
//     i64; floats are not produced by the kit/mint flow).
//   - Strings: UTF-8 verbatim, escape `"` and `\\` and U+0000..U+001F as
//     `\u00XX` (lowercase hex). RFC 8785 also permits the named short
//     escapes (\n etc.) but the C++/Rust peers chose `\u00XX` for
//     determinism; we match.
//   - true / false / null verbatim.
//   - No whitespace anywhere.
//
// The Value tree is intentionally tiny. We keep insertion order on
// objects so callers can build envelopes naturally; the JCS encoder
// re-sorts keys at emit time.

const std = @import("std");

pub const ValueKind = enum {
    null_,
    bool_,
    integer,
    string,
    array,
    object,
};

/// Owned (key, value) pair for an Object Value. Values are reference-counted-
/// like via pointer ownership: the `Value.deinit()` recursively frees children.
pub const Pair = struct {
    key: []u8,
    value: *Value,
};

pub const Value = union(ValueKind) {
    null_: void,
    bool_: bool,
    integer: i64,
    string: []u8,
    array: []*Value,
    object: []Pair,

    /// Recursively free this Value and all children allocated through `alloc`.
    /// `alloc` MUST be the same allocator that built the tree.
    pub fn deinit(self: *Value, alloc: std.mem.Allocator) void {
        switch (self.*) {
            .null_, .bool_, .integer => {},
            .string => |s| alloc.free(s),
            .array => |items| {
                for (items) |child| {
                    child.deinit(alloc);
                    alloc.destroy(child);
                }
                alloc.free(items);
            },
            .object => |entries| {
                for (entries) |pair| {
                    alloc.free(pair.key);
                    pair.value.deinit(alloc);
                    alloc.destroy(pair.value);
                }
                alloc.free(entries);
            },
        }
    }

    // ---- constructors -----------------------------------------------------

    pub fn newNull(alloc: std.mem.Allocator) !*Value {
        const v = try alloc.create(Value);
        v.* = .{ .null_ = {} };
        return v;
    }

    pub fn newBool(alloc: std.mem.Allocator, b: bool) !*Value {
        const v = try alloc.create(Value);
        v.* = .{ .bool_ = b };
        return v;
    }

    pub fn newInt(alloc: std.mem.Allocator, n: i64) !*Value {
        const v = try alloc.create(Value);
        v.* = .{ .integer = n };
        return v;
    }

    pub fn newString(alloc: std.mem.Allocator, s: []const u8) !*Value {
        const owned = try alloc.dupe(u8, s);
        const v = try alloc.create(Value);
        v.* = .{ .string = owned };
        return v;
    }

    /// Build an Array node taking ownership of `items` (slice of *Value).
    /// `items` must have been allocated through the same `alloc`.
    pub fn newArrayOwned(alloc: std.mem.Allocator, items: []*Value) !*Value {
        const v = try alloc.create(Value);
        v.* = .{ .array = items };
        return v;
    }

    /// Build an Object node taking ownership of `entries`. Keys are
    /// expected to already be `alloc.dupe`'d.
    pub fn newObjectOwned(alloc: std.mem.Allocator, entries: []Pair) !*Value {
        const v = try alloc.create(Value);
        v.* = .{ .object = entries };
        return v;
    }
};

/// Object builder: collects (key, value) pairs in insertion order, then
/// transfers ownership to a freshly-built Object Value. Keys are dupe'd
/// on `add`; values are taken by pointer (the builder does not copy them).
pub const ObjectBuilder = struct {
    alloc: std.mem.Allocator,
    pairs: std.ArrayList(Pair) = .empty,

    pub fn init(alloc: std.mem.Allocator) ObjectBuilder {
        return .{ .alloc = alloc };
    }

    pub fn add(self: *ObjectBuilder, key: []const u8, value: *Value) !void {
        const owned_key = try self.alloc.dupe(u8, key);
        try self.pairs.append(self.alloc, .{ .key = owned_key, .value = value });
    }

    /// Finalize. The returned *Value owns the entries slice and all keys/
    /// child Values transitively.
    pub fn finish(self: *ObjectBuilder) !*Value {
        const entries = try self.pairs.toOwnedSlice(self.alloc);
        return Value.newObjectOwned(self.alloc, entries);
    }
};

/// Array builder: same shape as ObjectBuilder.
pub const ArrayBuilder = struct {
    alloc: std.mem.Allocator,
    items: std.ArrayList(*Value) = .empty,

    pub fn init(alloc: std.mem.Allocator) ArrayBuilder {
        return .{ .alloc = alloc };
    }

    pub fn append(self: *ArrayBuilder, v: *Value) !void {
        try self.items.append(self.alloc, v);
    }

    pub fn finish(self: *ArrayBuilder) !*Value {
        const slice = try self.items.toOwnedSlice(self.alloc);
        return Value.newArrayOwned(self.alloc, slice);
    }
};

// ---------------------------------------------------------------------------
// Encoder
// ---------------------------------------------------------------------------

/// Encode `v` as JCS-canonical JSON. Returned slice is owned by the caller.
pub fn encode(alloc: std.mem.Allocator, v: *const Value) ![]u8 {
    var buf: std.ArrayList(u8) = .empty;
    defer buf.deinit(alloc);
    try encodeValue(alloc, &buf, v);
    return buf.toOwnedSlice(alloc);
}

fn encodeValue(alloc: std.mem.Allocator, out: *std.ArrayList(u8), v: *const Value) !void {
    switch (v.*) {
        .null_ => try out.appendSlice(alloc, "null"),
        .bool_ => |b| try out.appendSlice(alloc, if (b) "true" else "false"),
        .integer => |n| {
            // ECMA-262 ToString applied to a finite integer; std.fmt's `{d}`
            // emits the same form (no leading zeros, leading minus for
            // negatives, no trailing decimal).
            try out.print(alloc, "{d}", .{n});
        },
        .string => |s| try encodeString(alloc, out, s),
        .array => |items| {
            try out.append(alloc, '[');
            for (items, 0..) |child, i| {
                if (i > 0) try out.append(alloc, ',');
                try encodeValue(alloc, out, child);
            }
            try out.append(alloc, ']');
        },
        .object => |entries| {
            // Sort by key (byte-order works for ASCII keys; the protocol's
            // keys are all ASCII). For non-ASCII keys this is still the
            // right answer because UTF-8 byte order matches Unicode
            // code-point order on well-formed input.
            const sorted = try alloc.alloc(Pair, entries.len);
            defer alloc.free(sorted);
            @memcpy(sorted, entries);
            std.mem.sort(Pair, sorted, {}, pairLess);

            try out.append(alloc, '{');
            for (sorted, 0..) |pair, i| {
                if (i > 0) try out.append(alloc, ',');
                try encodeString(alloc, out, pair.key);
                try out.append(alloc, ':');
                try encodeValue(alloc, out, pair.value);
            }
            try out.append(alloc, '}');
        },
    }
}

fn pairLess(_: void, a: Pair, b: Pair) bool {
    return std.mem.lessThan(u8, a.key, b.key);
}

fn encodeString(alloc: std.mem.Allocator, out: *std.ArrayList(u8), s: []const u8) !void {
    try out.append(alloc, '"');
    // Iterate by byte: non-ASCII bytes (>= 0x80) are emitted verbatim,
    // preserving the input's UTF-8 encoding. The Rust peer iterates by
    // Unicode scalar; the result is identical for well-formed UTF-8 input
    // because we only special-case ASCII-range characters.
    for (s) |b| {
        if (b == '"') {
            try out.appendSlice(alloc, "\\\"");
        } else if (b == '\\') {
            try out.appendSlice(alloc, "\\\\");
        } else if (b < 0x20) {
            const hex = "0123456789abcdef";
            try out.appendSlice(alloc, "\\u00");
            try out.append(alloc, hex[(b >> 4) & 0xF]);
            try out.append(alloc, hex[b & 0xF]);
        } else {
            try out.append(alloc, b);
        }
    }
    try out.append(alloc, '"');
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

test "encode_simple_object_sorts_keys" {
    const alloc = std.testing.allocator;
    var ob = ObjectBuilder.init(alloc);
    try ob.add("b", try Value.newInt(alloc, 1));
    try ob.add("a", try Value.newString(alloc, "x"));
    const v = try ob.finish();
    defer {
        v.deinit(alloc);
        alloc.destroy(v);
    }
    const enc = try encode(alloc, v);
    defer alloc.free(enc);
    try std.testing.expectEqualStrings("{\"a\":\"x\",\"b\":1}", enc);
}

test "encode_nested_array_object" {
    const alloc = std.testing.allocator;
    var arr = ArrayBuilder.init(alloc);
    try arr.append(try Value.newInt(alloc, 1));
    try arr.append(try Value.newInt(alloc, 2));
    var ob = ObjectBuilder.init(alloc);
    try ob.add("xs", try arr.finish());
    const v = try ob.finish();
    defer {
        v.deinit(alloc);
        alloc.destroy(v);
    }
    const enc = try encode(alloc, v);
    defer alloc.free(enc);
    try std.testing.expectEqualStrings("{\"xs\":[1,2]}", enc);
}

test "escape_quotes_and_backslash" {
    const alloc = std.testing.allocator;
    const v = try Value.newString(alloc, "a\"b\\c");
    defer {
        v.deinit(alloc);
        alloc.destroy(v);
    }
    const enc = try encode(alloc, v);
    defer alloc.free(enc);
    try std.testing.expectEqualStrings("\"a\\\"b\\\\c\"", enc);
}

test "empty_object_and_array" {
    const alloc = std.testing.allocator;
    {
        var ob = ObjectBuilder.init(alloc);
        const v = try ob.finish();
        defer {
            v.deinit(alloc);
            alloc.destroy(v);
        }
        const enc = try encode(alloc, v);
        defer alloc.free(enc);
        try std.testing.expectEqualStrings("{}", enc);
    }
    {
        var arr = ArrayBuilder.init(alloc);
        const v = try arr.finish();
        defer {
            v.deinit(alloc);
            alloc.destroy(v);
        }
        const enc = try encode(alloc, v);
        defer alloc.free(enc);
        try std.testing.expectEqualStrings("[]", enc);
    }
}

test "unicode_atomic_predicates_round_trip_verbatim" {
    // Regression: cross-language hash agreement depends on this. The kit's
    // atomic predicate names use exactly these (>=, <=, !=). The Rust JCS
    // emitter preserves these chars verbatim (UTF-8 bytes >= 0x80 emitted
    // as-is); we match.
    const alloc = std.testing.allocator;
    const inputs = [_][]const u8{ "\u{2265}", "\u{2264}", "\u{2260}" };
    for (inputs) |sym| {
        const v = try Value.newString(alloc, sym);
        defer {
            v.deinit(alloc);
            alloc.destroy(v);
        }
        const enc = try encode(alloc, v);
        defer alloc.free(enc);
        // Encoded form is `"<sym>"`: the same UTF-8 bytes wrapped in quotes.
        try std.testing.expectEqual(@as(usize, sym.len + 2), enc.len);
        try std.testing.expectEqual(@as(u8, '"'), enc[0]);
        try std.testing.expectEqual(@as(u8, '"'), enc[enc.len - 1]);
        try std.testing.expectEqualSlices(u8, sym, enc[1 .. enc.len - 1]);
    }
}

test "negative_integer_serializes_with_leading_minus" {
    const alloc = std.testing.allocator;
    const v = try Value.newInt(alloc, -42);
    defer {
        v.deinit(alloc);
        alloc.destroy(v);
    }
    const enc = try encode(alloc, v);
    defer alloc.free(enc);
    try std.testing.expectEqualStrings("-42", enc);
}

test "control_char_escaped_lowercase_hex" {
    const alloc = std.testing.allocator;
    const v = try Value.newString(alloc, "\x1f");
    defer {
        v.deinit(alloc);
        alloc.destroy(v);
    }
    const enc = try encode(alloc, v);
    defer alloc.free(enc);
    try std.testing.expectEqualStrings("\"\\u001f\"", enc);
}

test "object_key_sort_is_byte_order" {
    // Per RFC 8785 §3.2.3: sort by Unicode code-point. For ASCII keys this
    // is byte-order; "Z" (0x5A) sorts before "a" (0x61).
    const alloc = std.testing.allocator;
    var ob = ObjectBuilder.init(alloc);
    try ob.add("a", try Value.newInt(alloc, 1));
    try ob.add("Z", try Value.newInt(alloc, 2));
    const v = try ob.finish();
    defer {
        v.deinit(alloc);
        alloc.destroy(v);
    }
    const enc = try encode(alloc, v);
    defer alloc.free(enc);
    try std.testing.expectEqualStrings("{\"Z\":2,\"a\":1}", enc);
}
